text/0081-truncate.md

```diff
@@ -0,0 +1,105 @@
+
+- there is a network isolation between primary and DR DC
+- PD make the replica state of primary DC to be Async after a while
+- the primary DC is down
+
+Then the sync state of the cluster (from the DR DC's PD) is not dependable. In this case, data can be partially sent to the TiKV nodes in DR, and break the ACID consistency.
+
+Though it is impossible to do lossless recovery in this situation, we can recover the data on TiKV nodes to a certain version (`resolved_ts` in this case), which is ACID consistency.
+
+## Detailed design
+
+Basically, we need a method to return TiKV to a certain version (timestamp), we should add a new RPC to kvproto:
+
+```protobuf
+message TruncateRequest {
+    Context context = 1;
+    uint64 checkpoint_ts = 2;
+}
+
+message TruncateResponse {  
+    string error = 1;
+}
+
+service Tikv {
+    // …
+    rpc Truncate(kvrpcpb.TruncateRequest) returns (kvrpcpb.TruncateResponse) {}
+}
+```
+
+When a TiKV receive `TruncateRequest`, it will create a `TruncateWorker` instance and start the work.
+
+There are several steps to follow when doing truncate, note it is expected that no more write requests will be sent to TiKV during the truncate process:
+
+### 1. Wait all committed raft log are applied
+
+We can just polling each region's `RegionInfo` in a method similar with [`Debugger::region_info`](https://github.com/tikv/tikv/blob/789c99666f2f9faaa6c6e5b021ac0cf7a76ae24e/src/server/debug.rs#L188) does, and wait for `apply_index` to be equal to `commit_index`.
+
+### 2. Clean WriteCF and DefaultCF
+
+Then we can remove all data created by transactions which committed after `checkpoint_ts`, this can be done by remove all `Write` records which `commit_ts > checkpoint_ts`  and corresponding data in `DEFAULT_CF` from KVEngine, ie.
+
+```rust
+fn scan_next_batch(&mut self, batch_size: usize) -> Option<Vec<(Vec<u8>, Write)>> {
+    let mut writes = None;
+    for _ in 0..batch_size {
+        if let Some((key, write)) = self.next_write()? {
+            let commit_ts = Key::decode_ts_from(keys::origin_key(&key));
+            let mut writes = writes.get_or_insert(Vec::new());
+            if commit_ts > self.ts {
+                writes.push((key, write));
+            }
+        } else {
+            return writes;
+        }
+    }
+    writes
+}
+
+pub fn process_next_batch(
+    &mut self, 
+    batch_size: usize,
+    wb: &mut RocksWriteBatch,
+) -> bool {
+    let writes = if let Some(writes) = self.scan_next_batch(batch_size) {
+        writes
+  	} else {
+        return false;
+  	}
+    for (key, write) in writes {
+        let default_key = Key::append_ts(Key::from_raw(&key), write.start_ts).to_raw().unwrap();
+        box_try!(wb.delete_cf(CF_WRITE, &key));
+        box_try!(wb.delete_cf(CF_DEFAULT, &default_key));
+    }
+  	true
+}
+```
+
+Note we scan and process data in batches, so we can report the progress to the client.
+
+### 3. Resolve locks once
```

Comment:
Is it guaranteed that `checkpoint_ts` must be some resolved_ts? I guess it's true, otherwise resolving these locks may lead to new `Write` whose commit_ts > checkpoint_ts (async commit).

If that's true, why not just remove all the locks in step 2?